tla-pp/src/main/scala/at/forsyte/apalache/tla/pp/SetMembershipSimplifier.scala

```diff
@@ -27,46 +36,55 @@ class SetMembershipSimplifier(tracker: TransformationTracker) extends AbstractTr
-   * *not* applicable to `Nat`, since `i \in Nat` does not hold for all `IntT1`-typed `i`.
+   * Returns true iff rewriting of a well-typed set-membership test `x \in arg` to `TRUE` is applicable for the
+   * function's argument.
+   *
+   * The applicable sets are inductively defined as
+   *   - the predefined sets BOOLEAN, Int, Real, STRING,
+   *   - sets of sequences over applicable sets, e.g., Seq(BOOLEAN), Seq(Int), Seq(Seq(Int)), Seq(SUBSET Int), ...
+   *   - power sets of applicable sets, e.g., SUBSET BOOLEAN, SUBSET Int, SUBSET Seq(Int), ...
+   *   - sets of functions over applicable sets, e.g., [Int -> BOOLEAN], ...
+   *
+   * In particular, it is *not* applicable to `Nat` and sequence sets / power sets thereof, since `i \in Nat` does not
+   * hold for all `IntT1`-typed `i`.
    */
-  private def typeOfSupportedPredefSet: PartialFunction[TlaPredefSet, TlaType1] = {
-    case TlaBoolSet => BoolT1()
-    case TlaIntSet  => IntT1()
-    case TlaRealSet => RealT1()
-    case TlaStrSet  => StrT1()
-    // intentionally omits TlaNatSet, see above.
-  }
+  private def isApplicable: Function[TlaEx, Boolean] = {
+    // base case: BOOLEAN, Int, Real, STRING
+    case ValEx(TlaBoolSet) | ValEx(TlaIntSet) | ValEx(TlaRealSet) | ValEx(TlaStrSet) => true
 
-  /**
-   * Checks if this transformation is applicable (see [[typeOfSupportedPredefSet]]) to a TLA+ predefined set `ps` of
-   * primitives, and if the types of `name` and `ps` match.
-   */
-  private def isApplicable(name: TlaEx, ps: TlaPredefSet): Boolean =
-    typeOfSupportedPredefSet.isDefinedAt(ps) && name.typeTag == Typed(typeOfSupportedPredefSet(ps))
+    // inductive cases:
+    // 1. Seq(s) for applicable set `s`
+    case OperEx(TlaSetOper.seqSet, set) => isApplicable(set)
+    // 2. SUBSET s for applicable set `s`
+    case OperEx(TlaSetOper.powerset, set) => isApplicable(set)
+    // 3. [s1 -> s2] for applicable sets `s1` and `s2
+    case OperEx(TlaSetOper.funSet, set1, set2) => isApplicable(set1) && isApplicable(set2)
 
-  /**
-   * Checks if this transformation is applicable (see [[typeOfSupportedPredefSet]]) to a TLA+ predefined set of
-   * sequences (`Seq(_)`) `ps`, and if the types of `name` and `ps` match.
-   */
-  private def isApplicableSeq(name: TlaEx, ps: TlaPredefSet): Boolean =
-    typeOfSupportedPredefSet.isDefinedAt(ps) && name.typeTag == Typed(SeqT1(typeOfSupportedPredefSet(ps)))
+    // otherwise
+    case _ => false
+  }
 
   /**
-   * Rewrites vacuously true membership tests based on type information, and rewrites `i \in Nat` to `i \ge 0`.
+   * Simplifies expressions commonly found in `TypeOK`, assuming they are well-typed.
    *
-   * For example, `x \in BOOLEAN` is rewritten to `TRUE` if `x` is typed `BoolT1`.
+   * @see
+   *   [[SetMembershipSimplifier]] for a full list of supported rewritings.
    */
   private def transformMembership: PartialFunction[TlaEx, TlaEx] = {
```

Comment:
Since all the cases have to be `OperEx( TlaSetOper.in, name, ...)` you could do
```scala
private def transformMembership: TlaEx => TlaEx = {
  case OperEx(TlaSetOper.in, name, setEx) => 
    setEx match {
      // x \in AS  ~>  TRUE
      case set if isApplicable(set) => trueVal
      // n \in Nat  ~>  x >= 0
      case ValEx(TlaNatSet) => OperEx(TlaArithOper.ge, name, ValEx(TlaInt(0))(intTag))(boolTag)
      // fun \in [Dom -> AS]  ~>  DOMAIN fun = Dom    (fun \in [AS1 -> AS2] is handled above)
      case OperEx(TlaSetOper.funSet, domain, set2) if isApplicable(set2) =>
        OperEx(TlaOper.eq, OperEx(TlaFunOper.domain, name)(domain.typeTag), domain)(boolTag)
      case _ => ex
    }
  case _ => ex
}
```

I'm also not too keen on the use of partial functions; they make it harder to track down bugs, because the code will just skip cases where the partial function isn't applicable, instead of complaining that all cases aren't covered by the transformation (which may or may not be an implementation error).

Reply:
Good idea with the refactoring!

Regarding the partial function, I'm overriding `AbstractTransformer.partialTransformers` above which is a `Seq[PartialFunction[TlaEx, TlaEx]]` and `Function1` isn't a subtype of `PartialFunction`. I suppose there's nothing to be done here, unless you have an idea how to work around this?

Reply:
Then keep it as Partial I suppose, lets not derail this PR.
But @konnov, I propose to make `AbstractTransformer` use actual functions. Using `PartialFunctions` fells like it's supporting laziness, not efficiency.

Reply:
Refactoring in [9387f09](https://github.com/informalsystems/apalache/pull/1443/commits/9387f0937111e7c2e63b83531c3aea090ed1dec2)